lib/spack/spack/spec.py

```diff
@@ -2615,12 +2615,17 @@ def patches(self):
         if 'patches' not in self.variants:
             return []
 
+        variants_patches = self.variants['patches']
+
+        if not hasattr(variants_patches, '_patches_in_order_of_appearance'):
```

Comment:
why was this needed?  Are you seeing specs without it?

Reply:
Yes, sometimes I was getting an exception without this (I believe it was when I was testing creating a buildcache entry after I made the change which added the `full_hash`).  Perhaps it was just something about the state of my repo and local configs at the time?  At any rate, sometimes `variants_patches` did not have a `_patches_in_order_of_appearance` attribute, so I just guarded it.

Reply:
@tgamblin I found some notes where I had documented (kind of) the issue that was happening here.  It happened when creating a buildcache entry:

```
==> creating binary cache file for package python@2.7.15%gcc@5.4.0+dbm~optimizations patches=123082ab3483ded78e86d7c809e98a804b3465b4683c96bd79a2fd799f572244 +pic+pythoncmd+shared~tk~ucs4 arch=linux-ubuntu16.04-x86_64
==> Error: 'MultiValuedVariant' object has no attribute '_patches_in_order_of_appearance'
```

I have checked out the current `develop` branch, and tried creating a spec for python similar to the one shown above, then concretizing and `full_hash`-ing it, but I could not reproduce the issue.

I wonder if the thing to do is to remove the change from my PR until such time as I encounter the issue again?  Let me know what you think.